proto/api/v1beta1/ensign.proto

```diff
@@ -85,28 +109,76 @@ message Ack {
 // should return a Nack if the event couldn't be handled or processed so that Ensign
 // ensures another client retrieves the event.
 message Nack {
-    string id = 1;
-    uint32 code = 2;
+    enum Code {
+        UNKNOWN = 0;
+
+        // Server-side NACK codes.
+        MAX_EVENT_SIZE_EXCEEDED = 1;
+        TOPIC_UKNOWN = 2;
+        TOPIC_ARCHVIVED = 3;
+        TOPIC_DELETED = 4;
+        PERMISSION_DENIED = 5;
+        CONSENSUS_FAILURE = 6;
+        SHARDING_FAILURE = 7;
+        REDIRECT = 8;
+        INTERNAL = 9;
+
+        // Client-side NACK codes
+        UNPROCESSED = 100;
+        TIMEOUT = 101;
+        UNHANDLED_MIMETYPE = 102;
+        UNKNOWN_TYPE = 103;
+        DELIVER_AGAIN_ANY = 104;
+        DELIVER_AGAIN_NOT_ME = 105;
+    }
+
+    bytes id = 1;
+    Code code = 2;
     string error = 3;
 }
 
-// OpenStream is the first message that should be sent in a Subscribe stream as it
-// defines what topic the client is subscribing to and how it will process the events.
+// OpenStream is the first message that should be sent in a Publish stream in order to
+// identify and authenticate the publisher. Optionally, the publisher can specify the
+// topics that they want to publish to in order to check that they are being sent to the
+// correct node.
 message OpenStream {
-    string consumer_id = 1;
+    string client_id = 1;
```

Comment:
So the `client_id` is just generated by the client and has nothing to do with the API key?

Reply:
That's correct - because multiple clients could be using the same API key.